### PR TITLE
Humble desired pose check

### PR DESCRIPTION
**Pull request by Renbago**
## Basic Info

this pr include desiredPoseChecker service in slam_toolbox_localization, and few more karto_debug

---

## What's added ?

This service basicly using the same logic with 2d pose. But changing the angle offset,dimension's, min_distance, and few more settings. We are triggering karto for desired solution with desired settings. Service file returns best_response reloceted_x reloceted_y for best Solution. We are able to access those data with ```GetBestResponse``` function

## service type:
```'slam_toolbox/desired_pose_check'```
## referance service call
```ros2 service call /SR5S1/slam_toolbox/desired_pose_check slam_toolbox/srv/DesiredPoseChecker "{pose_x: 0.7, pose_y: 0.26, do_relocalization: true, search_distance: 12.0}"```

## Description of documentation updates required from your changes
those parameters need to add slam_toolbox.yaml
```
position_search_distance: 16.0   
position_search_resolution: 0.05
position_search_smear_deviation: 0.03
position_search_fine_angle_offset: 0.00349
position_search_coarse_angle_offset: 3.14
position_search_coarse_angle_resolution: 0.0349
position_search_do_relocalization: true
```
---

## Future 
?

### CODE REVIEW

**Comment by Renbago** [comment]
Currently do not accept this version. Its still test version, I'll be change things from mapper and I might be need changes few more. @oguzhankose @Zarqu0n

**Comment by Zarqu0n** [comment]
Is this okay to merge? @Renbago

**Comment by Renbago** [comment]
> Is this okay to merge? @Renbago 

Not yet, couldnt test in zorlu and I need to edit few things, due time 01 october

File: lib/karto_sdk/src/Mapper.cpp [hunk 1]
```diff
@@ -764,6 +787,24 @@ kt_double ScanMatcher::CorrelateScan(
   }
   assert(math::DoubleEqual(m_yPoses.back(), -startY));
 
+  // mx ve my poses  benim grid size ımdaki resamplelanmış pose ları tekabul ediyor
```
Comment (Zarqu0n) [inline lib/karto_sdk/src/Mapper.cpp]: is it important?

File: lib/karto_sdk/src/Mapper.cpp [hunk 2]
```diff
@@ -812,12 +863,19 @@ kt_double ScanMatcher::CorrelateScan(
     }
   }
 
+  end = std::chrono::steady_clock::now();
```
Comment (Zarqu0n) [inline lib/karto_sdk/src/Mapper.cpp]: is this publish message when robot normal working?

File: lib/karto_sdk/src/Mapper.cpp [hunk 3]
```diff
@@ -1529,10 +1598,10 @@ kt_bool MapperGraph::TryCloseLoop(LocalizedRangeScan * pScan, const Name & rSens
 
     std::chrono::time_point<std::chrono::system_clock> start, end;  
     start = std::chrono::system_clock::now();
-
 #endif
     Pose2 bestPose;
     Matrix3 covariance;
+    std::cout << "-----------------the next scan match is from trycloseloop------------" << std::endl;
```
Comment (Zarqu0n) [inline lib/karto_sdk/src/Mapper.cpp]: is this publish message when robot normal working?

File: lib/karto_sdk/src/Mapper.cpp [hunk 4]
```diff
@@ -2667,16 +2810,19 @@ void Mapper::setParamAngleVariancePenalty(double d)
 void Mapper::setParamFineSearchAngleOffset(double d)
 {
   m_pFineSearchAngleOffset->SetValue((kt_double)d);
+  std::cout << "m_pFineSearchAngleOffset: " << *m_pFineSearchAngleOffset << std::endl;
```
Comment (Zarqu0n) [inline lib/karto_sdk/src/Mapper.cpp]: is this publish message when robot normal working?

File: lib/karto_sdk/src/Mapper.cpp
Comment (Zarqu0n) [inline lib/karto_sdk/src/Mapper.cpp:1]: Please make sure that the robot does not log when operating normally. You can test this to disable navigation.

File: src/slam_toolbox_common.cpp
Comment (Zarqu0n) [inline src/slam_toolbox_common.cpp:1]: I thing we need to test any garbage variable. Maybe some variable were forgotten.

**Review by Zarqu0n — CHANGES_REQUESTED** [review]

File: src/slam_toolbox_common.cpp
Comment (Renbago) [reply src/slam_toolbox_common.cpp:1]: yeah i know i will clean those, beacuse of that i wrote do not merge.

**Comment by Renbago** [comment]
@Zarqu0n review

**Review by Zarqu0n — COMMENTED** [review]
Did you test in robot? Has publish any useless log message.

**Comment by Renbago** [comment]
> Did you test in robot? Has publish any useless log message.

yes but i've found bug do not merge.

**Comment by Renbago** [comment]
I'ts mergable now.  @Zarqu0n

**Comment by Renbago** [comment]
## If service called:
the terminal will gave info about called service with relocalize situation.
terminal outputs are:
```
RCLCPP_INFO(get_logger(), "LocalizationSlamToolbox: Searching for best response with relocalize");
or 
RCLCPP_INFO(get_logger(), "LocalizationSlamToolbox: Searching for best response without relocalize");
```

the reference service response with call:
```
ros2 service call /SR5S1/slam_toolbox/desired_pose_check slam_toolbox/srv/DesiredPoseChecker "{pose_x: 0.7, pose_y: 10.26, do_relocalization: true, search_distance: 8.
0}"                                                                                                                                                                                        
                                                                                                                                                                                                
requester: making request: slam_toolbox.srv.DesiredPoseChecker_Request(pose_x=0.7, pose_y=10.26, search_distance=8.0, do_relocalization=True)                                              
                                                                                                                                                                                           
response:                                                                                                                                                                                  
slam_toolbox.srv.DesiredPoseChecker_Response(best_response=0.6311222222222223, relocated_x=0.53, relocated_y=10.279999732971191, relocated_yaw=0.0, message='Found bestResponse', success=True)                                                                                                                                                                                       
                                                                                                                                                                                           
```

I also write in [wiki](https://wiki.saharobotik.com/en/autonomy/Localization/new-page) so for who wants to access,

**Pull request merged**